Tentative mana spreader path tracing optimizations

quat1024 opened the pull request:
Mana spreader code is lovely as we all know so I'd like to get a few more eyes on this, if at all possible? (And someone who knows how to use a profiler would be really nice, ha!)

This PR:
* caps the maximum mana burst beam-scan distance to 400 ticks into the future, preventing big lags from https://github.com/Vazkii/Botania/issues/2787
  * This changes gameplay behavior but nobody will actually care
* stops beam-scan mana bursts from searching for entities (this information was simply thrown away, resulting in lots and lots and lots of wasted calls to `getEntitiesWithinAABBExcludingEntity`)
  * The Tripwire lens still works btw
* stops beam-scan mana bursts from sending "pings" to the mana spreader - this information was also simply thrown away, resulting in a lot of useless calls to `getTileEntity`
  * This is something I am unsure about, maybe something breaks here.
* comments out "inWater" related code from all mana bursts - I was never able to make a mana burst actually appear `inWater`, no matter how I aimed it at a pool of water.

Potential future areas of spreader/burst optimization (read: maybe don't merge yet?):

* is PositionProperties a hot spot? lots of little allocations there I'm sure
* what about `getLensInstance`?
* `onEntityUpdate` seems to be doing a *huge* amount of useless things. Running particles? Riding entities? Fire damage? DEFINITELY check this out. Lots of stuff that's not really that bad on its own but when you have a bunch of mana bursts doing these pointless checks every tick + spreaders occasionally doing 400 or so in a single tick... it can add up

Feel free to look in to any of those areas - I'll probably do it myself later when I get the time, but it's like 2:30am

williewillus commented:
1. I would recommend against overriding things randomly to noop them unless it produces a real, measurable impact. It's more to maintain and keep track of when porting
2. If I recall correctly, the data manager already stores things in deserialized form internally, so this wouldn't help much
3. Yes, the simulation time should be configurable I think.

Hubry commented:
Profiling definitely is a good idea and Quat actually wanted to do that before doing more in this area (and I had to remind him of the profiler again).

Anyways, I did a little bit of profiling with Spark with the latest Botania release. Small testing setup of: 3 mana spreaders with resistance lenses pointing directly into a chain of 7 resistance prisms, with 7 blocks of space before a wall. 
For client thread profiling, held a wand of the forest and did `/sparkclient start --include-line-numbers --timeout 60 --interval 1` while looking at the setup.
For server thread profiling, set up a few pistons moving blocks just before the wall on a 3 second timer and did `/spark start --include-line-numbers --timeout 360 --interval 1`.

[Client results](https://sparkprofiler.github.io/#Zn9AiHZL5z), [server results](https://sparkprofiler.github.io/#f9sklj71Yt).
Seems like checking if the burst is in lava, entity search as well as all data manager calls combined are somewhat heavy (seems like data manager might be the heaviest thing when all the calls are added), while PositionProperties has a negligible impact. I'd say no-oping some of these things I mentioned might be a viable option.

quat1024 commented:
Vazkii likes it so we're going to merge it pretty much as-is, hope that's okey:eyes:

I added a configuration option for the spreader trace time (there's no min or max value on it, I can trust people to be sensible here :stuck_out_tongue: ).

I also nooped `isInLava` since it's a pretty free fix for a relatively expensive function (according to Hubry's profiling about as bad as the pointless `getEntitiesWithinAABBExcludingEntity` checks, ouch)

> I would recommend against overriding things randomly to noop them unless it produces a real, measurable impact. It's more to maintain and keep track of when porting

Good point - I added a comment to the `isInLava` override to hopefully explain why it's there and I'll be more careful about this type of thing ifwhen I revisit optimizing.
